src/browser/history.js

```diff
@@ -0,0 +1,321 @@
+
+  Tag.remove = (tag, item) =>
+    tag.update('items', exclude(item[PouchDB.id]));
+
+  Tag.tag = (item, tagName) => item.update('tags', include(tagName));
+
+  Tag.untag = (item, tagName) => item.update('tags', exclude(tagName));
+
+  const Top = Record({
+    [PouchDB.id]: ID('top/pages'),
+    [PouchDB.revision]: Maybe(Revision),
+    pages: List(Page)
+  });
+  Top.storeID = Page.storeID;
+  Top.sample = (page, limit) => top => top.update('pages', pages => {
+    const index = pages.findIndex(x => x[PouchDB.id] === page[PouchDB.id])
+    return pages.set(index < 0 ? pages.size : index, page)
+                .sortBy(Page.frequency)
+                .take(limit);
+  });
+
+  const clear = async(function*({stores}) {
+    yield stores.pages.destroy();
+    yield stores.quotes.destroy();
+    yield stores.tags.destroy();
+  });
+
+  let taskID = 0;
+
+  const scheduleEdit = async(function*(history, record, transform) {
+    try {
+      // wait for last scehduled edit to complete.
+      yield history.editQueue[record[PouchDB.id]];
+    } catch (error) {
+      // If last operation faild log an error, but proceed to a next
+      // operation.
+      console.error(error);
+    } finally {
+      const store = history.stores[record.constructor.storeID];
+      return edit(record, transform, store);
+    }
+  })
+
+  // History
+  class History {
+    static defaults() {
+      return {
+        pagesStoreName: "pages",
+        quotesStoreName: "quotes",
+        tagsStoreName: "tags",
+        topPageLimit: 6,
+        trackTopPages: false
+      }
+    }
+    constructor(options={}) {
+      this.onPageChange = this.onPageChange.bind(this);
+      this.onTopPagesChange = this.onTopPagesChange.bind(this);
+
+      this.options = Object.assign(History.defaults(), options);
+      const {pagesStore, quotesStore, tagsStore,
+             pagesStoreName, quotesStoreName, tagsStoreName} = this.options;
+
+      this.stores = {
+        pages: pagesStore || new PouchDB(pagesStoreName),
+        quotes: quotesStore || new PouchDB(quotesStoreName),
+        tags: tagsStore || new PouchDB(tagsStoreName)
+      }
+
+      this.editQueue = Object.create(null);
+
+      if (options.trackTopPages) {
+        this.setupChangeFeeds();
+        this.setupListeners();
+      }
+    }
+    setupChangeFeeds() {
+      const {pages} = this.stores
+
+      this.topPageChangeFeed = pages.changes({
+        since: "now",
+        live: true,
+        include_docs: true,
+        doc_ids: ["top/pages"]
+      });
+
+      this.pagesChangeFeed = pages.changes({
+        since: "now",
+        live: true,
+        filter: ({[PouchDB.id]: id}) => id.startsWith("page/"),
+        include_docs: true
+      });
+    }
+    setupListeners() {
+      this.topPageChangeFeed.on("change", this.onTopPagesChange);
+      this.pagesChangeFeed.on("change", this.onPageChange);
+    }
+
+    // Edits per record are queued, to avoid data loss
+    // due to concurrent edits.
```

Comment:
Anyway I don't mind adding retry a.k.a "upsert" on conflicts althugh we still should  keep the queing to avoid wasting resources.